aws/resource_aws_autoscaling_group.go

```diff
@@ -1369,6 +1447,148 @@ func resourceAwsAutoscalingGroupDelete(d *schema.ResourceData, meta interface{})
+				if callerr, ok := err.(awserr.Error); ok {
+					switch callerr.Code() {
+					case "ResourceInUse", "ScalingActivityInProgress":
+						// These are retryable
+						return resource.RetryableError(callerr)
+					}
+				}
+				// Didn't recognize the error, so shouldn't retry.
+				return resource.NonRetryableError(err)
+			}
+			// Successful delete
+			return nil
+		})
+
+		if isResourceTimeoutError(err) {
+			_, err = conn.DeleteWarmPool(&deleteopts)
+		}
+		if err != nil {
+			return fmt.Errorf("Error deleting Warm Pool: %s", err)
+		}
+	}
+
+	// Wait for Warm pool to be gone.
+	err := resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
+		g, _ := getAwsAutoscalingGroup(d.Id(), conn)
+
+		if g.WarmPoolConfiguration != nil {
+			return resource.RetryableError(fmt.Errorf("Warm pool still exists"))
+		}
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		g, err = getAwsAutoscalingGroup(d.Id(), conn)
+
+		if err != nil || g.WarmPoolConfiguration != nil {
+			return fmt.Errorf("Warm pool still exists")
+		}
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error deleting Warm Pool: %s", err)
+	}
+
+	log.Printf("[INFO] Successfully removed Warm pool")
+
+	return nil
+}
+
+func getAwsAutoscalingGroupWarmPool(asgName string, conn *autoscaling.AutoScaling) (*autoscaling.DescribeWarmPoolOutput, error) {
+	describeOpts := autoscaling.DescribeWarmPoolInput{
+		AutoScalingGroupName: aws.String(asgName),
+	}
+
+	log.Printf("[DEBUG] Warm Pool describe configuration input: %#v", describeOpts)
+	describeWarmPoolOutput, err := conn.DescribeWarmPool(&describeOpts)
+	if err != nil {
+		autoscalingerr, ok := err.(awserr.Error)
+		if ok && autoscalingerr.Code() == "InvalidGroup.NotFound" {
+			return nil, nil
+		}
+
+		return nil, fmt.Errorf("Error retrieving Warm Pool: %s", err)
+	}
+
+	return describeWarmPoolOutput, nil
+}
+
+func resourceAwsAutoscalingGroupWarmPoolDrain(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).autoscalingconn
+
+	if d.Get("force_delete").(bool) || d.Get("force_delete_warm_pool").(bool) {
+		log.Printf("[DEBUG] Skipping Warm pool drain, force delete was set.")
+		return nil
+	}
+
+	// First, set the max group prepared capacity and min size to zero for the pool to drain
+	log.Printf("[DEBUG] Reducing Warm pool capacity to zero")
+	opts := autoscaling.PutWarmPoolInput{
+		AutoScalingGroupName:     aws.String(d.Id()),
+		MaxGroupPreparedCapacity: aws.Int64(0),
+		MinSize:                  aws.Int64(0),
+	}
+	if _, err := conn.PutWarmPool(&opts); err != nil {
+		return fmt.Errorf("Error setting capacity to zero to drain: %s", err)
+	}
+
+	// Next, wait for the Warm Pool to drain
+	log.Printf("[DEBUG] Waiting for warm pool to have zero instances")
+	var p *autoscaling.DescribeWarmPoolOutput
+	err := resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
+		p, err := getAwsAutoscalingGroupWarmPool(d.Id(), conn)
+		if err != nil {
```

Comment:
Are there specific retryable errors that could be checked for so that if an unexpected error is returned, it doesn't get stuck in the retry loop for the full timeout?

Reply:
This is a describe API, there should not be any transient state / retryable exceptions from the service.

Are you thinking of exceptions like throttling exceptions here?

Reply:
Sorry, I misread this one. The way you've got it looks good.